deployer/gateway_name_deployer_test.go

```diff
@@ -0,0 +1,702 @@
+	ctrl := gomock.NewController(t)
+
+	defer ctrl.Finish()
+
+	identity, err := substrate.NewIdentityFromEd25519Phrase("")
+	assert.NoError(t, err)
+	deployer := mock.NewMockDeployer(ctrl)
+	sub := mock.NewMockSubstrateExt(ctrl)
+	gatewayNameDeployer := NewGatewayNameDeployer(&TFPluginClient{
+		Identity:      identity,
+		SubstrateConn: sub,
+		TwinID:        11,
+	})
+	gatewayNameDeployer.deployer = deployer
+	gw := workloads.GatewayNameProxy{
+		ContractID:       100,
+		NodeID:           10,
+		NodeDeploymentID: map[uint32]uint64{10: 100},
+		NameContractID:   200,
+		Name:             "name",
+		TLSPassthrough:   false,
+		Backends:         []zos.Backend{"https://1.1.1.1", "http://2.2.2.2"},
+		FQDN:             "name.com",
+	}
+	dls, err := gatewayNameDeployer.GenerateVersionlessDeployments(context.Background(), &gw)
+	assert.NoError(t, err)
+	dl := dls[10]
+	dl.Workloads[0].Result.State = gridtypes.StateOk
+	dl.Workloads[0].Result.Data, err = json.Marshal(zos.GatewayProxyResult{FQDN: "name.com"})
+	assert.NoError(t, err)
+	sub.EXPECT().DeleteInvalidContracts(
+		gw.NodeDeploymentID,
+	).Return(nil)
+	sub.EXPECT().IsValidContract(
+		gw.NameContractID,
+	).Return(true, nil)
+
+	deployer.EXPECT().
+		GetDeployments(gomock.Any(), map[uint32]uint64{10: 100}).
+		Return(map[uint32]gridtypes.Deployment{10: dl}, nil)
+	gw.FQDN = "123"
+	err = gatewayNameDeployer.sync(context.Background(), &gw)
+	assert.Equal(t, gw.FQDN, "name.com")
+	assert.NoError(t, err)
+	assert.Equal(t, gw.NodeDeploymentID, map[uint32]uint64{10: 100})
+	assert.Equal(t, gw.NameContractID, uint64(200))
+	assert.Equal(t, gw.ContractID, uint64(100))
+}
+
+func TestNameSyncDeletedWorkload(t *testing.T) {
+	ctrl := gomock.NewController(t)
+
+	defer ctrl.Finish()
+
+	identity, err := substrate.NewIdentityFromEd25519Phrase("")
+	assert.NoError(t, err)
+	deployer := mock.NewMockDeployer(ctrl)
+	sub := mock.NewMockSubstrateExt(ctrl)
+	gatewayNameDeployer := NewGatewayNameDeployer(&TFPluginClient{
+		Identity:      identity,
+		SubstrateConn: sub,
+		TwinID:        11,
+	})
+	gatewayNameDeployer.deployer = deployer
+	gw := workloads.GatewayNameProxy{
+		ContractID:       100,
+		NodeID:           10,
+		NodeDeploymentID: map[uint32]uint64{10: 100},
+		NameContractID:   200,
+		Name:             "name",
+		TLSPassthrough:   false,
+		Backends:         []zos.Backend{"https://1.1.1.1", "http://2.2.2.2"},
+		FQDN:             "name.com",
+	}
+	dls, err := gatewayNameDeployer.GenerateVersionlessDeployments(context.Background(), &gw)
+	assert.NoError(t, err)
+	dl := dls[10]
+	// state is deleted
+
+	sub.EXPECT().DeleteInvalidContracts(
+		gw.NodeDeploymentID,
+	).Return(nil)
+	sub.EXPECT().IsValidContract(
+		gw.NameContractID,
+	).Return(true, nil)
+
+	deployer.EXPECT().
+		GetDeployments(gomock.Any(), map[uint32]uint64{10: 100}).
+		Return(map[uint32]gridtypes.Deployment{10: dl}, nil)
+	gw.FQDN = "123"
+	err = gatewayNameDeployer.sync(context.Background(), &gw)
+	assert.NoError(t, err)
```

Comment:
we should check that the contract exists.
`assert.Equal(t, gw.ContractID, uint64(100))`
`assert.Equal(t, gw.NodeDeploymentID, map[uint32]uint64{10: 100})`